### PR TITLE
C implement address anti malleability checksum check.

### LINKED ISSUE #220: C impl address malleability checksum.
Encoding is prone to malleability. So base64, base58 etc. are easily malleable. To avoid that we are introducing a checksum into the public part of the encoded public address. 
https://github.com/bartossh/Computantis/blob/2d5585d2182fca53412b7a17d86517de0005f38e/wallet/wallet.go#L106

The same is required for C implementation.

### PR DESCRIPTION
**Pull request by bartossh**
Any encoding is prone to malleability. So base64, base58 etc. are easily malleable. To avoid that we are introducing a checksum into the public part of the encoded public address. 
https://github.com/bartossh/Computantis/blob/2d5585d2182fca53412b7a17d86517de0005f38e/wallet/wallet.go#L106

The same is required for C implementation. 

Based on today's discussion with @kubagruszka @dmatusiewicz-consult-red @kmroz and me, there where a fair argument about checksum validation in any reliable data transfer (given by @dmatusiewicz-consult-red ).  

So the C needs anti-malleability checksum validation based on sha256. This is the implementation. Go already has it.

How it works:
1. This is the RAW BUFFER to base58 encode address: [ 1 byte of wallet version, ...32 bytes of public address ...,  checksum sha256 last 4 bytes]

2. Encode (version + public key +  RAW BUFFER )-> base56 encode -> "base58 string representing address".

3. It is almost impossible to corrupt the base58 encoded address and still produce valid public key. We need to satisfy the valid wallet version and the checksum.
4. We are using the last 4 bytes only in the base58 encoded address just because it is computationally very, very expensive to brute force the generation of a new valid public address that agrees on the version and the last 4  bytes of a checksum is correct. The probability is near zero, lower than the one calculated by Edward Taller for Oppenheimer regarding the probability of an atomic bomb activating hydrogen atoms. If they were satisfied I think we can. Another groaner joke... forgive me.

### CODE REVIEW

File: c/client-node/address/address.c [hunk 1]
```diff
@@ -40,22 +89,64 @@ char *encode_address_from_raw(unsigned char *raw, size_t len)
...
     }
-    
-    bool ok = b58tobin(*raw, &len, str, strlen(str));
-    if (!ok)
+
+    flag = memcpy(*raw, decoded + 1, PUBLIC_KEY_LEN);
+    if (flag == NULL)
     {
-        printf("Base58 decoding faild\n");
+        printf("Copying public key failed failed\n");
```
Comment (kmroz) [inline c/client-node/address/address.c]: nit: double "failed"

File: c/client-node/address/address.c [hunk 2]
```diff
@@ -40,22 +89,64 @@ char *encode_address_from_raw(unsigned char *raw, size_t len)
...
+        exit(1);
+    }
+
+    unsigned char pub_key_vrs[1+PUBLIC_KEY_LEN];
+    pub_key_vrs[0] = version;
+    flag = memcpy(pub_key_vrs+1, *raw, PUBLIC_KEY_LEN);
+    if (flag == NULL)
+    {
+        printf("Copying public key failed failed\n");
```
Comment (kmroz) [inline c/client-node/address/address.c]: nit: double "failed"

File: c/client-node/address/address.h [hunk 3]
```diff
@@ -10,7 +10,8 @@
 #ifndef ADDRESS_H
 #define ADDRESS_H
 #define PUBLIC_KEY_LEN 32
-#define ADDRESS_LEN 51
+#define ADDRESS_LEN 56
+#define CHECKSUM_LEN 4
 
 #include <ctype.h>
 #include <stdbool.h>
```
Comment (kmroz) [inline c/client-node/address/address.h:17]: size_t is defined in <stddef.h>.  Depending on header ordering in the calling source file, we can end up with a compilation failure.  Perhaps add <stddef.h> here?

File: c/client-node/address/address.c [hunk 4]
```diff
@@ -40,22 +89,64 @@ char *encode_address_from_raw(unsigned char *raw, size_t len)
...
         exit(1);
     }
 
+    unsigned char target_checksum[CHECKSUM_LEN];
+    checksum(pub_key_vrs, target_checksum, 1+PUBLIC_KEY_LEN, CHECKSUM_LEN);
+
+    int equality = strncmp((char *)actual_checksum, (char *)target_checksum, CHECKSUM_LEN);
```
Comment (kmroz) [inline c/client-node/address/address.c:147]: request: could you perhaps add a small note as to why they should differ?  Or better yet, what do you think about removing the version from the checksum altogether?  Then actual_checksum and target_checksum should be equal?  Or am I not understanding this correctly?

File: c/client-node/address/address.h [hunk 5]
```diff
@@ -20,7 +21,7 @@
 /// Encoder uses base58 encoding algorithm and returns pointer to nullable string.
 /// Caller takes responsibility of freeing the returned string. 
 ///
-char *encode_address_from_raw(unsigned char *raw, size_t len);
+char *encode_address_from_raw(unsigned char wallet_version, unsigned char  *raw, size_t len);
 
 /// 
 /// decode_address_to_raw decodes nullable string to raw bytes.
```
Comment (kmroz) [inline c/client-node/address/address.h]: nit: I like the doc string on `encode_address_from_raw`.  This is the mirror operation, so perhaps something like:

`decode_address_to_raw decodes a public address string into a raw buffer/bytes array.  It returns the number of bytes in the raw buffer, or 0 on error.  On a successful invocation, the caller is responsible to free the unsigned char *raw bytes array.`

Something along those lines.  What do you think?

**Review by kmroz — CHANGES_REQUESTED** [review]
I think solid!  Few small comments.  Let me know what you think.

File: c/client-node/address/address.c (hunk 4 above)
Comment (bartossh) [reply c/client-node/address/address.c:147]: It is a bug!!! Greate catch, I made a mistake somewhere... 🤦 ...

**Review by kmroz — APPROVED** [review]

**Pull request merged**